eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs

```diff
@@ -279,22 +279,26 @@ epochTransition = do
   let
     pulsingState = epochState0 ^. epochStateDRepPulsingStateL
 
-    ratState0@RatifyState {rsRemoved, rsEnactState} = extractDRepPulsingState pulsingState
+    ratState0@RatifyState {rsRemoved, rsEnacted, rsEnactState} = extractDRepPulsingState pulsingState
 
     (accountState2, dState2, newEnactState) =
       applyEnactedWithdrawals accountState1 dState1 rsEnactState
-    (newProposals, removedGovActions) =
-      -- It is important that we use current proposals here instead of the ones from the pulser
-      proposalsRemoveIds rsRemoved (govState0 ^. cgProposalsL)
+
+    -- It is important that we use current proposals here instead of the ones from the pulser
+    (proposals0, removedInvalidGovActions) = proposalsRemoveDescendentIds rsRemoved (govState0 ^. cgProposalsL)
+    (proposals1, removedEnactedGovActions) = proposalsRemoveIds rsEnacted proposals0
```

Comment:
@lehins Regarding the fix for #4000 I think this is the site at which we can notice that enacted actions weren't being cleaned up at all before this PR was merged. Only `rsRemoved` (expired) actions were being removed before, and without clean up of their descendents.